File: formula-java/src/main/java/org/formula/parse/tree/OptionalNodeExpression.java
package org.formula.parse.tree;

import java.util.List;

public class OptionalNodeExpression implements NodeExpression {
    private final List<NodeExpression> expressions;

    public static OptionalNodeExpression of(List<NodeExpression> expressions) {
        return new OptionalNodeExpression(expressions);
    }

    @Override
    public <T> Node<T> chainTo(Node<T> root) {
        Node<T> node = root;
        for (NodeExpression toAdd : expressions) {
            node = toAdd.chainTo(node);
        }
        if (root instanceof MappableNode<T> mappableRoot && node instanceof MappableNode<T> mappableNode) {
            mappableRoot.mapper(token -> {
                TokenMapper<T> mapper = mappableNode.mapper();
                if (mapper == null) {
                    return null;
                }
                return mapper.map(token);
            });
        }
        return node;
    }

    private OptionalNodeExpression(List<NodeExpression> expressions) {
        this.expressions = expressions;
    }
}

File: cs/formula-cs/Formula/TokenTree/OptionalNodeExpression.cs
namespace Formula.TokenTree;

internal class OptionalNodeExpression : INodeExpression
{
    private readonly IEnumerable<INodeExpression> _expressions;

    public static OptionalNodeExpression Of(IEnumerable<INodeExpression> expressions) {
        return new OptionalNodeExpression(expressions);
    }
    
    public Node<T> ChainTo<T>(Node<T> root)
    {
        var node = root;
        foreach (var toAdd in _expressions) {
            node = toAdd.ChainTo(node);
        }
        if (root is MappableNode<T> mappableRoot && node is MappableNode<T> mappableNode)
        {
            mappableRoot.Mapper = token =>
            {
                var mapper = mappableNode.Mapper;
                return mapper == null ? default : mapper(token);
            };
        }
        return node;
    }
    
    private OptionalNodeExpression(IEnumerable<INodeExpression> expressions)
    {
        _expressions = expressions;
    }
}